transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamRequestManager.java

```diff
@@ -0,0 +1,133 @@
+ */
+package io.camunda.zeebe.transport.stream.impl;
+
+import io.atomix.cluster.MemberId;
+import io.atomix.cluster.messaging.ClusterCommunicationService;
+import io.camunda.zeebe.scheduler.ConcurrencyControl;
+import io.camunda.zeebe.transport.stream.impl.messages.AddStreamRequest;
+import io.camunda.zeebe.transport.stream.impl.messages.RemoveStreamRequest;
+import io.camunda.zeebe.transport.stream.impl.messages.StreamTopics;
+import io.camunda.zeebe.util.buffer.BufferReader;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Handles sending add/remove stream request to the servers. */
+final class ClientStreamRequestManager<M extends BufferWriter, P extends BufferReader> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ClientStreamRequestManager.class);
+  private static final byte[] REMOVE_ALL_REQUEST = new byte[0];
+  private static final Duration RETRY_DELAY = Duration.ofSeconds(1);
+  private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);
+  private final ClusterCommunicationService communicationService;
+  private final ConcurrencyControl executor;
+
+  ClientStreamRequestManager(
+      final ClusterCommunicationService communicationService, final ConcurrencyControl executor) {
+    this.communicationService = communicationService;
+    this.executor = executor;
+  }
+
+  void openStream(final ClientStream<M, P> clientStream, final Collection<MemberId> servers) {
+    final var request =
+        new AddStreamRequest()
+            .streamType(clientStream.getStreamType())
+            .streamId(clientStream.getStreamId())
+            .metadata(clientStream.getMetadata());
+
+    servers.forEach(brokerId -> executor.run(() -> doAdd(request, brokerId, clientStream)));
+  }
+
+  private void doAdd(
+      final AddStreamRequest request,
+      final MemberId brokerId,
+      final ClientStream<M, P> clientStream) {
+    if (clientStream.isConnected(brokerId)) {
+      return;
+    }
+
+    final CompletableFuture<byte[]> result =
+        communicationService.send(
+            StreamTopics.ADD.topic(),
+            request,
+            BufferUtil::bufferAsArray,
+            Function.identity(),
+            brokerId,
+            REQUEST_TIMEOUT);
+
+    result.whenComplete(
+        (ignored, error) -> {
+          if (error != null) {
+            LOG.warn(
+                "Failed to open stream {} to node {}. Will retry in {}",
+                clientStream,
+                brokerId,
+                RETRY_DELAY);
+            // TODO: define some abort conditions. We may not have to retry indefinitely.
+            // For now we retry always. Eventually the request will succeed. Duplicate add request
```

Comment:
@npepinpe I remove the "optimization" we added in the prototype to not retry if it was already acknowledged. It might prevent sending a new request to a restarted node because the previous request is still marked as acknowledged. As mentioned in the PR requests, we have to re-visit retry handling.